### PR TITLE
mirror: Support mirroring to disk as a target

### PR DESCRIPTION
**Pull request by smarterclayton**
Allows release mirroring to disk. `oc adm release mirror` is updated to use
this mechanism. The mirror command, when --to-dir is specified, prints out
the command that can be used to restore the content.

Mirroring a release to disk:

    oc adm release mirror quay.io/openshift-release-dev/ocp-release:4.1.18 --to-dir=/tmp/release

Mirroring that back to another registry

    oc image mirror --from-dir=/tmp/release file://openshift/release myregistry.com/my/repository

Now that mirror supports local file destinations, add a simple server command under `oc image serve` that allows someone to mirror a release to disk and then host it within a firewall.

I.e.:

    oc adm release mirror ... --to-dir=/srv/repo

    # Move inside the firewall

    oc image serve --dir=/srv/repo

The server performs minimal HTTP header mangling to support modern container registry clients. Optional --tls-key and --tls-cert flags allow the server to listen on TLS.

### CODE REVIEW

**Comment by smarterclayton** [comment]
/assign @abhinavdahiya 

Gradually improving the flexibility of the commands to allow offline content replication.  Open to suggestions (I'm trying to improve the flexibility of the `image` commands while making release mirroring easier).  Builds on feedback from folks trying to clone things offline.

I think this would be a candidate for back port to simplify 4.2 offline support.

**Comment by abhinavdahiya** [comment]
I like this, will try to review by EOD.

**Comment by smarterclayton** [comment]
Specifically this is different than skopeo because

a) it creates a disk layout that matches the HTTP layout of a registry server, like we already do for s3
b) the disk layout matches the human assumptions about the registry
c) it allows us to reuse blobs (although there is an optimization missing which would scan the directory for duplicate blobs and link them during mirror)
d) this will be useful in ci and testing flows to recreate specific repos
e) this is the most barebones possible registry approach and is fairly tractable to keep working

File: pkg/cli/admin/release/mirror.go [hunk 1]
```diff
@@ -525,17 +501,23 @@ func (o *MirrorOptions) Run() error {
...
+		}
 	}
-
-	if o.PrintImageContentInstructions {
-		if err := printImageContentInstructions(o.Out, o.From, o.To, repositories); err != nil {
-			return fmt.Errorf("Error creating mirror usage instructions: %v", err)
+	if len(o.ToDir) > 0 {
+		fmt.Fprintf(o.Out, "\nTo upload local images to a registry, run:\n\n    oc image mirror --from-dir=%s file://%s* REGISTRY/REPOSITORY\n\n", o.ToDir, to)
+	} else if len(o.To) > 0 {
+		if o.PrintImageContentInstructions {
```
Comment (abhinavdahiya) [inline pkg/cli/admin/release/mirror.go]: should we move this to conditional block https://github.com/openshift/oc/pull/126/commits/a8c5d8f8c5cce6c81675e360d37f42fffeb3b9e1#diff-6c5f7c1cbfa2abdaa2847f0982a31e5fR507 ??

File: pkg/cli/image/mirror/file.go [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+	return distribution.Descriptor{
+		Digest: godigest.Digest(filepath.Base(target)),
+		Size:   fi.Size(),
+	}, nil
+}
+
+// Tag associates the tag with the provided descriptor, updating the
+// current association, if needed.
+func (s *fileTagStore) Tag(ctx context.Context, tag string, desc distribution.Descriptor) error {
+	panic("not implemented")
```
Comment (abhinavdahiya) [inline pkg/cli/image/mirror/file.go]: nit: I would rather see an error, or a little description for why panic is most appropriate in the commit message.

File: pkg/cli/image/mirror/file.go [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+// Untag removes the given tag association
+func (s *fileTagStore) Untag(ctx context.Context, tag string) error {
+	panic("not implemented")
```
Comment (abhinavdahiya) [inline pkg/cli/image/mirror/file.go]: same as https://github.com/openshift/oc/pull/126/commits/a8c5d8f8c5cce6c81675e360d37f42fffeb3b9e1#r334689057

File: pkg/cli/image/mirror/file.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+		if fi.Mode()&os.ModeSymlink != 0 {
+			names = append(names, filepath.Base(fi.Name()))
+		}
+	}
+	return names, nil
+}
+
+// Lookup returns the set of tags referencing the given digest.
+func (s *fileTagStore) Lookup(ctx context.Context, digest distribution.Descriptor) ([]string, error) {
+	panic("not implemented")
```
Comment (abhinavdahiya) [inline pkg/cli/image/mirror/file.go]: same as https://github.com/openshift/oc/pull/126/commits/a8c5d8f8c5cce6c81675e360d37f42fffeb3b9e1#r334689057

File: pkg/cli/image/mirror/file.go [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "manifests", dgst.String())
+	fi, err := os.Lstat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if fi.IsDir() || fi.Mode()&os.ModeSymlink != 0 {
+		return false, fmt.Errorf("not a file")
```
Comment (abhinavdahiya) [inline pkg/cli/image/mirror/file.go]: include the path in the message maybe? `<path> is not a file`

File: pkg/cli/image/mirror/file.go [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+	}
+	klog.V(5).Infof("Read manifest %T from %s: %v", manifest, path, desc)
+	return manifest, nil
+}
+
+func atomicWrite(path string, data []byte) error {
+	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(path+".download", data, 0600); err != nil {
```
Comment (abhinavdahiya) [inline pkg/cli/image/mirror/file.go]: is it expected that the path can never be a directory path?

File: pkg/cli/image/mirror/file.go [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+		}
+	}
+	return s.r.newWriter(path, d.String(), opts.Mount.Stat.Size), nil
+}
+
+func (s *fileBlobStore) Resume(ctx context.Context, id string) (distribution.BlobWriter, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// fileWriter attempts to upload parts to S3 in a buffered fashion where the last
```
Comment (abhinavdahiya) [inline pkg/cli/image/mirror/file.go]: doesn't seem like this is doing anything related to s3.

**Comment by abhinavdahiya** [comment]
> https://github.com/openshift/oc/pull/126/commits/a8c5d8f8c5cce6c81675e360d37f42fffeb3b9e1

The `release/mirror.go` changed look good. 
I don't have complete grasp of `cli/mirror` changes, went through them as a single pass and had few questions.. will keep digging on those. But for sanity check we should have somebody from experienced with image-registry stufff take a peek.

**Comment by pearj** [comment]
I just discovered if the pull secret file doesn't exist it segfaults:
ie running: `./oc adm -a pull_secret_doesnexist.json release mirror quay.io/openshift-release-dev/ocp-release:4.2.0 --to-dir=/tmp/ocp-release`
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x1fafc59]

goroutine 1 [running]:
github.com/openshift/oc/pkg/cli/image/manifest.(*SecurityOptions).Context(0xc00037f130, 0x7ffe89fd5816, 0x15, 0xc0011c09a0)
        /root/oc/pkg/cli/image/manifest/manifest.go:90 +0x59
github.com/openshift/oc/pkg/cli/image/extract.(*Options).Run(0xc00037f0e0, 0x0, 0x0)
        /root/oc/pkg/cli/image/extract/extract.go:306 +0x4a
github.com/openshift/oc/pkg/cli/admin/release.(*ExtractOptions).Run(0xc0001e1dc0, 0xc000518860, 0x2af1f05)
        /root/oc/pkg/cli/admin/release/extract.go:210 +0x3eb
github.com/openshift/oc/pkg/cli/admin/release.(*MirrorOptions).Run(0xc001259040, 0x0, 0x2cd1c30)
        /root/oc/pkg/cli/admin/release/mirror.go:276 +0x2aea
github.com/openshift/oc/pkg/cli/admin/release.NewMirror.func1(0xc00049d400, 0xc000e22c80, 0x1, 0x4)
        /root/oc/pkg/cli/admin/release/mirror.go:80 +0xa0
github.com/spf13/cobra.(*Command).execute(0xc00049d400, 0xc000e22c40, 0x4, 0x4, 0xc00049d400, 0xc000e22c40)
        /root/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x2ae
github.com/spf13/cobra.(*Command).ExecuteC(0xc001078a00, 0xb, 0xc001078a00, 0xb)
        /root/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fc
github.com/spf13/cobra.(*Command).Execute(...)
        /root/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
        /root/oc/cmd/oc/oc.go:103 +0x815
```

File: pkg/cli/image/mirror/file.go (hunk 7 above)
Comment (smarterclayton) [reply pkg/cli/image/mirror/file.go]: copy-pasta

File: pkg/cli/image/mirror/file.go (hunk 6 above)
Comment (smarterclayton) [reply pkg/cli/image/mirror/file.go]: yes, in that there is only one level of file leaves allowed (no dir in this structure that has files has subduers

**Comment by smarterclayton** [comment]
Thanks, separate bug but will update it here

**Comment by smarterclayton** [comment]
Comments addressed, did a bit more manual testing.

**Comment by smarterclayton** [comment]
I want to get some soak time on this, and get e2es going.  Going to tag today unless there are other comments.

**Pull request merged**